File: apps/ios/ScribbleFit/Sources/Features/Ledger/Data/Repository/LedgerRepositoryImpl.swift
import Foundation

@MainActor
public final class LedgerRepositoryImpl: LedgerRepository {
    private let database: ScribbleFitDatabase

    public init(database: ScribbleFitDatabase) {
        self.database = database
    }

    public func getWorkoutHistory() async throws -> [WorkoutHistory] {
        let logs = database.getAllWorkoutLogs()
        return logs.map { log in
            let sets = database.getSetsForWorkout(id: log.id)
            let grouped = Dictionary(grouping: sets, by: \.exerciseId)
            let exercises = grouped.map { exerciseId, exerciseSets in
                ExerciseHistory(
                    canonicalName: exerciseId,
                    sets: exerciseSets.map { s in
                        SetHistory(weight: s.weight, reps: s.reps, rpe: s.rpe, notes: s.notes)
                    }
                )
            }
            return WorkoutHistory(
                id: log.id,
                date: log.date,
                location: log.location,
                totalVolume: log.totalVolume ?? 0,
                exercises: exercises
            )
        }
    }

    public func logWorkout(_ workout: WorkoutHistory) async throws {
        let exercises = workout.exercises.map { exercise in
            ExerciseEntity(
                id: exercise.canonicalName,
                canonicalName: exercise.canonicalName,
                muscleGroup: "",
                aliases: []
            )
        }
        database.insertExercisesIfAbsent(exercises)

        database.upsertWorkoutLog(WorkoutEntity(
            id: workout.id,
            date: workout.date,
            location: workout.location,
            totalVolume: workout.totalVolume
        ))

        let sets = workout.exercises.flatMap { exercise in
            exercise.sets.map { set in
                WorkoutSet(
                    id: UUID().uuidString,
                    weight: set.weight,
                    reps: set.reps,
                    rpe: set.rpe,
                    notes: set.notes,
                    exerciseId: exercise.canonicalName
                )
            }
        }
        database.upsertWorkoutSets(sets)
    }
}

File: apps/android/feature/ledger/src/main/java/com/scribblefit/feature/ledger/data/repository/LedgerRepositoryImpl.kt
package com.scribblefit.feature.ledger.data.repository

import com.scribblefit.core.database.dao.ExerciseDao
import com.scribblefit.core.database.dao.SetDao
import com.scribblefit.core.database.dao.WorkoutDao
import com.scribblefit.feature.workout.domain.Exercise
import com.scribblefit.feature.workout.domain.Set
import com.scribblefit.feature.workout.domain.Workout
import com.scribblefit.feature.ledger.domain.repository.LedgerRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class LedgerRepositoryImpl @Inject constructor(
    private val workoutDao: WorkoutDao,
    private val setDao: SetDao,
    private val exerciseDao: ExerciseDao
) : LedgerRepository {

    override fun getWorkoutHistory(): Flow<List<Workout>> =
        workoutDao.observeAll().map { logs ->
            logs.map { log ->
                val sets = setDao.getSetsForWorkout(log.id)
                val grouped = sets.groupBy { it.exerciseId }
                val exercises = grouped.map { (exerciseId, exerciseSets) ->
                    Exercise(
                        canonicalName = exerciseId,
                        muscleGroup = "",
                        sets = exerciseSets.map { s ->
                            Set(weight = s.weight, reps = s.reps, rpe = s.rpe, notes = s.notes)
                        }
                    )
                }
                Workout(
                    date = log.date,
                    exercises = exercises
                )
            }
        }

    override suspend fun logWorkout(exercise: Exercise) {

//        val exerciseEntities = workout.exercises.map { exercise ->
//            val entity = ExerciseEntity(
//                id = exercise.canonicalName,
//                canonicalName = exercise.canonicalName,
//                muscleGroup = "",
//                aliases = emptyList()
//            )
//        }
//
//        exerciseDao.insertExerciseIfAbsent(exerciseEntities)
//
//        workoutDao.upsert(
//            WorkoutEntity(
//                id = workout.id,
//                date = workout.date,
//                location = workout.location,
//                totalVolume = workout.totalVolume
//            )
//        )
//
//        val setEntities = workout.exercises.flatMap { exercise ->
//            exercise.sets.map { set ->
//                SetEntity(
//                    id = UUID.randomUUID().toString(),
//                    workoutId = workout.id,
//                    exerciseId = exercise.canonicalName,
//                    weight = set.weight,
//                    reps = set.reps,
//                    rpe = set.rpe,
//                    notes = set.notes
//                )
//            }
//        }
//        setDao.upsertAll(setEntities)
    }
}
